File: first_missing_positive.py
'''
https://leetcode.com/problems/first-missing-positive/
'''
def firstMissingPositive(nums: list[int]) -> int:
    '''
    Given an unsorted integer array nums, find the smallest missing positive integer.
    '''
    nums = sorted(set(filter(lambda x: x > 0, nums)))
    curr = 1
    for num in nums:
        if num > curr:
            return curr
        curr += 1
    return curr

if __name__ == '__main__':
    assert firstMissingPositive([7, 8, 9, 11, 12]) == 1
    assert firstMissingPositive([1, 2, 0]) == 3
    assert firstMissingPositive([3, 4, -1, 1]) == 2
    assert firstMissingPositive([1, 1, 1]) == 2

File: first_missing_positive.rs
//! <https://leetcode.com/problems/first-missing-positive/>
use std::collections::HashSet;
use std::iter::FromIterator;
pub fn first_missing_positive(nums: Vec<i32>) -> i32 {
    let max = nums.len() + 1;
    let set: HashSet<i32> = HashSet::from_iter(nums);
    for i in 1..max {
        if !set.contains(&(i as i32)) {
            return i as _;
        }
    }
    max as _
}

fn main() {
    assert_eq!(first_missing_positive(vec![1, 3, 6, 4, 1, 2]), 5);
    assert_eq!(first_missing_positive(vec![1, 2, 3]), 4);
    assert_eq!(first_missing_positive(vec![-1, -3]), 1);
}
